Badge listing UI

Kelketek opened the pull request:
**Description**: Provides a UI for viewing badges in a user's profile.
**Jira**: https://openedx.atlassian.net/browse/SOL-1365
**Dependencies**: https://github.com/edx/edx-platform/pull/10732
**Sandbox** http://pr10556.sandbox.opencraft.com/

edx-webhook commented:
Thanks for the pull request, @Kelketek! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?number=10861&repo=edx%2Fedx-platform

bradenmacdonald commented on lms/djangoapps/badges/migrations/0002_data__migrate_assertions.py:
```diff
@@ -43,14 +45,20 @@ def forwards(apps, schema_editor):
-        ).save()
+        )
+        assertion.save()
+        # Would be overwritten by the first save.
```
FYI, per Django docs, another option "If you want to be able to modify this field, set the following instead of `auto_now_add=True`: `default=timezone.now` - from `django.utils.timezone.now()`"

Kelketek replied:
@bradenmacdonald Yes, but aside from this initial use, it shouldn't need manual setting again.

Kelketek commented:
@bradenmacdonald I have this to the point where it displays the badges in the user's profile. To try it out, go log in as the honor user and view your profile page.

@talbs I'm reserving the share button and modal for a follow-up PR since I didn't get a chance to hit it today with all the fixes needed for this and the API. So I'd like to get the basic view layout looked over at your first opportunity while I work on the rest.

@nasthagiri This PR modifies the User account API slightly to allow the badges to show up in the profile. Please let me know if these modifications are acceptable. They're in https://github.com/edx/edx-platform/commit/44f3dcff4c5a15571f33e675e88bf39a5f206dbe

talbs commented:
@Kelketek is there a sandbox view that I can take a look at?

Kelketek commented:
@talbs http://pr10556.sandbox.opencraft.com/

talbs commented:
> @talbs http://pr10556.sandbox.opencraft.com/

Thanks much, @Kelketek. Is there a particular user on that sandbox that has earned badges or a way to get to the view you've built out with this work?

Kelketek commented:
@talbs honor@example.com, pw honor

talbs commented:
Nevermind, just found it under honor@example.com. Sorry!

talbs commented:
@Kelketek, thanks for the support in reviewing.

![profile badges-compare](https://cloud.githubusercontent.com/assets/163763/11691627/daef06e0-9e69-11e5-869a-d23943310104.png)

**^ Comparing the static comp/visual design with your rendering, there are a few things that need some adjustment:**

* The spacing between the new tabs you added and the profile image/details seems small.
* Font-size of the tab text seems small
* Font-sizes of the date/copy text of badges should be different
* Line-height of the details text seems less than what's expected
* The "Earn your next accomplishment" call to action area is missing the background/border/box-shadow treatment in the comp
* The empty badge "Earn your next accomplishment" dotted pattern is darker than expected and has a space at the top of it.
* Smaller things include: spacing between tab text and borders; vertical spacing between the two columns seems tight.

Hope that visual design feedback helps.

bradenmacdonald commented on lms/static/js/spec/student_profile/learner_profile_factory_spec.js:
```diff
@@ -62,6 +62,33 @@ define(['backbone', 'jquery', 'underscore', 'common/js/spec_helpers/ajax_helpers
                 LearnerProfileHelpers.expectLimitedProfileSectionsAndFieldsToBeRendered(learnerProfileView);
             });
 
+            it("doesn't render the mode toggle if there are no badges", function() {
```
Is this a requirement? I don't see mention of it in UX-2980. IMHO it's confusing to hide this sometimes, and it means users with no badges won't see the "Earn Your Next Accomplishment" call to action.

bradenmacdonald commented on lms/static/js/student_profile/collections/badges_collection.js:
```diff
@@ -0,0 +1,14 @@
+        var BadgesCollection = Backbone.Collection.extend({
+            model : BadgesModel,
+            parse: function(response) {
+                console.log(response);
```
Make sure to remove this line when done this stage of development :)

bradenmacdonald commented on lms/static/js/student_profile/views/badge_listing_view.js:
```diff
@@ -0,0 +1,37 @@
+                    this.collection.each(function (badge, index, collection) {
+                        var keep_row = (index % 2);
+                        if (!keep_row) {
+                            row = $("<div class='badge-row'>");
```
I just tested and the HTML appears the same to me whether you put all columns ( > 2 ) inside a single "row" element or whether you manually split them up into rows. I'd advise just using one big "row" element and put all the badges into that single row, which is more flexible and simpler compared to manually assigning them into rows. Maybe just rename the class so it's not called a "row".

bradenmacdonald commented:
@Kelketek Not sure if this will ever happen, and you might have already fixed it, but: if someone uploads a rectangular badge image, the layout runs together and doesn't have enough padding:
![screen shot 2015-12-09 at 4 22 12 pm](https://cloud.githubusercontent.com/assets/945577/11703119/33a67e90-9e91-11e5-9e7a-68a6b5be6759.png)

Should be fixable by adding padding and/or margin to the `.badge-display`

Kelketek commented:
@bradenmacdonald That should not be possible as a rectangular image would be rejected normally.

@talbs This is ready for a second look from you.

@catong I don't think there's likely to be any significant changes to strings/error messages in this PR from this point forward, so you should be able to take a look at this now.

nasthagiri commented on lms/djangoapps/badges/models.py:
```diff
@@ -132,6 +132,7 @@ class BadgeAssertion(models.Model):
     backend = models.CharField(max_length=50)
     image_url = models.URLField()
     assertion_url = models.URLField()
+    created_at = models.DateTimeField(auto_now_add=True, db_index=True)
```
Consider using [TimeStampedModel](http://django-model-utils.readthedocs.org/en/latest/models.html#timestampedmodel) instead from `model_utils.models`.  You can see other uses of it in edx-platform.

nasthagiri commented on lms/static/js/spec/student_profile/helpers.js:
```diff
@@ -105,9 +105,78 @@ define(['underscore'], function(_) {
+                "course_id": null,
+                "description": "Yay! It's a test badge.",
+                "criteria": "https://example.com/syllabus",
+                "image": "http://localhost:8000/media/badge_classes/test_lMB9bRw.png"
```
"image" -> "image_url" per change from other PR.

nasthagiri replied:
plan to update this?

Kelketek replied:
@nasthagiri Looks like my change got clobbered by some git operation. I'll go fix it.

nasthagiri commented on lms/static/js/spec/student_account/helpers.js:
```diff
@@ -2,6 +2,7 @@ define(['underscore'], function(_) {
     'use strict';
 
     var USER_ACCOUNTS_API_URL = '/api/user/v0/accounts/student';
+    var BADGES_API_URL = '/api/user/v0/accounts/student/badges';
```
Shouldn't this be /api/badges/v1/...  ?

nasthagiri commented on openedx/core/djangoapps/user_api/accounts/serializers.py:
```diff
@@ -63,6 +65,12 @@ def to_representation(self, user):
         """
         profile = user.profile
+        badges = settings.FEATURES.get('ENABLE_OPENBADGES') or False
+        badges = badges and BadgeAssertionSerializer(
```
How large will this data get?  Should there be a parameter in the request that indicates whether or not to include the badges information?  A user's profile could be requested in many places.  For example, when get a list of profiles for all users in a discussion thread, we don't necessarily want all the details of each user.

catong commented on lms/templates/student_profile/badge_placeholder.underscore:
```diff
@@ -0,0 +1,10 @@
+    <span class="accomplishment-placeholder" aria-hidden="true">
+  </div>
+  <div class="badge-details">
+    <div class="badge-name"><%- gettext('Earn Your Next Accomplishment') %></div>
```
"What's Your Next Accomplishment?"

catong commented on lms/templates/student_profile/badge_placeholder.underscore:
```diff
@@ -0,0 +1,10 @@
+  </div>
+  <div class="badge-details">
+    <div class="badge-name"><%- gettext('Earn Your Next Accomplishment') %></div>
+    <p class="badge-description"><%- gettext('Find a course and start learning towards a goal.') %></p>
```
"Start working toward your next learning goal."
(Remove "Find a course" to avoid repeating the button text; and much as it hurts me to remove that "s", our style is the US "toward" rather than Cdn/UK "towards") :-)

talbs commented:
@Kelketek, things look generally good. One small thing, ``.badge-date-stamp`` should have a ``font-size`` of ``14px``, not ``12px``.

talbs commented:
@Kelketek, once you address that small visual change, :+1:

Kelketek commented:
@nasthagiri Ready for another look here.

Kelketek commented:
@catong Strings updated.

bradenmacdonald commented:
:+1: LGTM

nasthagiri commented on openedx/core/djangoapps/user_api/accounts/serializers.py:
```diff
@@ -63,6 +68,16 @@ def to_representation(self, user):
         profile = user.profile
+        badges = settings.FEATURES.get('ENABLE_OPENBADGES') or False
+        if 'badges' in self.excluded_fields:
+            # Avoid doing this potentially expensive query-- it will be filtered out anyway.
```
If the badges field is a potentially expensive query then why is it opt-out, instead of opt-in?  That is, why not have this be an "additional_fields" parameter instead of an "excluded_fields"?

Also, do you plan to paginate/cap this in a subsequent PR in case a user has a large number of badges?

Kelketek replied:
@nasthagiri I don't think it's likely to be all that expensive. It would make for larger JSON payloads, and it is possible that someone will have earned enough badges to cause a problem because it isn't paginated-- though unlikely, I'd think. It takes a while to earn a badge-- you can't just upload a bunch of them. There isn't a current plan to paginate in a follow-up PR.

However, I could remove this entirely from the User API, modify the Badges API to check if a user is sharing their profile fully, and display the badges if so. It and the subsequent PR won't be merged in time for Friday's release, however.

Kelketek replied:
In the current form, we're treating badges like a field on the profile. This API returns all public fields. Any field that shouldn't be returned should be opt out.

nasthagiri replied:
I strongly suggest we change this field to a boolean: 'has_accomplishments' for example.
This will allow the Profile UI to know whether to display the "accomplishments" tab.
And then have the UI call the badges API in a separate AJAX request.

This is exactly how we were planning to do it for mobile: boolean in Profile API, with paginated call to Badges API.
If you can do this, it would solve both Mobile's and Web's requirements, I believe.

nasthagiri replied:
By the way, is there a requirement to merge this before today's release cut?  It's not a requirement for mobile.  And I believe your team's Dec deadline is for merging to master, not release to production.

Kelketek replied:
@nasthagiri There is not a direct requirement, but there are a lot of stakeholders who would like to have these features and who would probably need to have them backported if they don't make it in.

Kelketek replied:
@talbs @antoviaque @nasthagiri The current UX compositions do not have anything for pagination. It will take a while to modify the code to allow this. I don't think we'll have it completed by this week, accordingly, though it should be possible to finish it next week.

nasthagiri replied:
If you change this to a Boolean in the profile API, doesn't it satisfy the Web UI's requirements?  You would then make a separate Ajax call to the badges API.

Sent from my iPhone

> On Dec 11, 2015, at 7:44 AM, Kelketek <notifications@github.com> wrote:
> 
> In openedx/core/djangoapps/user_api/accounts/serializers.py:
> 
> > @@ -63,6 +68,16 @@ def to_representation(self, user):
> >          :return: Dict serialized account
> >          """
> >          profile = user.profile
> > +        badges = settings.FEATURES.get('ENABLE_OPENBADGES') or False
> > +        if 'badges' in self.excluded_fields:
> > +            # Avoid doing this potentially expensive query-- it will be filtered out anyway.
> @talbs @antoviaque @nasthagiri The current UX compositions do not have anything for pagination. It will take a while to modify the code to allow this. I don't think we'll have it completed by this week, accordingly, though it should be possible to finish it next week.
> 
> —
> Reply to this email directly or view it on GitHub.
>

catong commented on lms/templates/student_profile/learner_profile.underscore:
```diff
@@ -13,6 +13,8 @@
             <i class="fa fa-exclamation-triangle message-error" aria-hidden=true></i>
             <span class="copy"><%- gettext("An error occurred. Please reload the page.") %></span>
```
@Kelketek Ugh, sorry I missed this: "An error occurred. Try loading the page again."

catong commented:
:+1: with last tweak to an error message that I didn't comment on in the first pass. Thanks!

Kelketek commented:
@pbaruah I've been informed that @cptvitamin is out sick. Is there someone else who can review this for accessibility?

antoviaque commented:
@Kelketek @pbaruah I don't think there is any one else who could review the accessibility. To try to get this in the release from today and in Dogwood, it might be worth deferring the accessibility review to a follow-up PR?

Is Mark's and @nasthagiri 's +1s the only ones you are missing to merge? @nedbat won't be reviewing this PR?

Kelketek commented:
@antoviaque I'm pretty sure @nedbat isn't intending to review this PR, no. He was originally listed on all the tickets but gave me the impression that he really only intended to review the core infrastructure changes and that the rest could be handled by the different teams.

catong commented:
@Kelketek I have checked with @clrux, also on the Accessibility team, and he can take a look. Chris, this is the Solutions PR I mentioned to you.

antoviaque commented:
@Kelketek Ok, sounds good. And Thanks @catong @clrux ! : )

clrux commented:
@Kelketek A few accessibility items I discovered:

The active tab (Accomplishments, etc.) should indicate audibly that it's active. We normally do this by using `.sr` text saying as much. Example: `<button class="toggle-button is-active" data-section=".badge-set-display">Accomplishments <span class="sr">(selected)</span></button>`

The badges themselves have no `alt` text, so the filename is read which is confusing. I recommend giving these badge graphics a name that can be used as the `alt` text. For example, "Course completed badge", "Batman Award", or "Course Fox". The idea being the announced text would describe what the badge is for. However, if that text is going to be repeated in the bolded first line of the description text, then the `alt` value for the badge graphics themselves should be left blank (`alt=""`).
<img width="577" alt="screen shot 2015-12-11 at 10 19 06 am" src="https://cloud.githubusercontent.com/assets/2112024/11747383/0023bfc2-9ff1-11e5-8142-e3a6aa22547c.png">


The badge course name is repeated and the date completed is confusing. I suggest breaking this up into digestible chunks. Imagine listening to this instead of seeing it. Something like:

"edX Demonstration Course"
"Completed with honors on 11-15-2015"
"Badge earned on 12-15-2015"
 
<img width="787" alt="screen shot 2015-12-11 at 10 19 29 am" src="https://cloud.githubusercontent.com/assets/2112024/11747474/68dd15f4-9ff1-11e5-94a8-c174e35ba435.png">


Additionally, the number of earned badges isn't present. I suggest making the badges list items with the total number of badges as a heading above. Example:

"Accomplishments"
"You have 10 accomplishment badges!"
[list badges here]

Which could be announced as:

"Accomplishments, selected button"
"You have ten accomplishment badges"
"List, with ten items"
"..."

nasthagiri commented on cms/djangoapps/models/settings/course_metadata.py:
```diff
@@ -50,7 +50,7 @@ class CourseMetadata(object):
         'is_time_limited',
         'is_practice_exam',
-        'self_paced'
+        'self_paced',
```
Nit: can revert this change.

nasthagiri commented on common/djangoapps/student/helpers.py:
```diff
@@ -12,6 +13,10 @@
 
 # Enumeration of per-course verification statuses
 # we display on the student dashboard.
+from openedx.core.djangoapps.user_api.accounts import PRIVATE_VISIBILITY, ACCOUNT_VISIBILITY_PREF_KEY, \
```
Nit: use parenthesis to include a list of imports instead of wrapping with a `\`.

nasthagiri commented on common/djangoapps/student/helpers.py:
```diff
@@ -230,3 +235,38 @@ def get_next_url_for_login_page(request):
     return redirect_to
+
+
+def get_profile_visibility(user_profile, user, configuration=None):
```
Why is this code moved here instead of exposed through something like `user_api.profiles.api.py`?
@cahrens Is there a recommended path for exposing user profile information to other django projects?

Kelketek replied:
I moved it here because it is where the origin data for user profiles is stored. So it seemed like the semantically correct place to put it. I could move it there instead.

Kelketek replied:
@nasthagiri Also, I tried moving it to where you suggested, but because the code there is now used in class construction for the Badging API view call permissions, it seems to cause a chicken-and-egg problem placing it where you suggest.

nasthagiri replied:
The placement of the code seems odd to me, but I'll defer to the T&L team, @cahrens, if they're Ok with it.

cahrens replied:
I would move it back where it came from. I definitely think it belongs in the user_api package.

nasthagiri commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -903,7 +903,7 @@ class CourseFields(object):
         ),
         scope=Scope.settings,
-        deprecated=True,  # Deprecated until the teams feature is made generally available
+        deprecated=False,  # Deprecated until the teams feature is made generally available
```
Why is this no longer deprecated?

Kelketek replied:
Diagnostic thing. Needs to be reverted. Thanks for catching that.

nasthagiri commented on lms/djangoapps/student_profile/views.py:
```diff
@@ -87,9 +85,14 @@ def learner_profile_context(request, profile_username, user_is_staff):
             'has_preferences_access': (logged_in_user.username == profile_username or user_is_staff),
             'own_profile': own_profile,
             'country_options': list(countries),
+            'find_courses_url': marketing_link('COURSES'),
```
Is this really needed?

Kelketek replied:
Yes. It is used to set the link on the 'Find a course' button, which may be different depending upon the instance.

nasthagiri replied:
Ok.

nasthagiri commented on lms/djangoapps/teams/__init__.py:
```diff
@@ -12,4 +12,6 @@ def is_feature_enabled(course):
     Returns True if the teams feature is enabled.
     """
+    print settings.FEATURES.get('ENABLE_TEAMS', False)
+    print course.teams_enabled
```
Please remove `print` statements.

nasthagiri commented on lms/djangoapps/teams/views.py:
```diff
@@ -128,10 +103,12 @@ def get(self, request, course_id):
 
         if not CourseEnrollment.is_enrolled(request.user, course.id) and \
                 not has_access(request.user, 'staff', course, course.id):
+            print "Not enrolled, or staff."
```
Reminder: more `print`s to remove in this file.

nasthagiri commented on lms/envs/common.py:
```diff
@@ -2596,6 +2596,8 @@
         'language_proficiencies',
         'bio',
         'account_privacy',
+        # Not an actual field, but used to signal whether badges should be public.
```
This is now an actual field, right?

Kelketek replied:
No. It is not stored in the user's preferences or on any model. It is purely configuration.

nasthagiri commented on openedx/core/djangoapps/user_api/accounts/api.py:
```diff
@@ -54,6 +54,7 @@ def get_account_settings(request, username=None, configuration=None, view=None):
         configuration (dict): an optional configuration specifying which fields in the account
             can be shared, and the default visibility settings. If not present, the setting value with
             key ACCOUNT_VISIBILITY_CONFIGURATION is used.
+        excluded (list): an optional list of fields not to return from the user's profile.
```
We no longer need the `excluded` parameter.

nasthagiri commented on openedx/core/djangoapps/user_api/accounts/serializers.py:
```diff
@@ -63,6 +63,7 @@ def to_representation(self, user):
         :return: Dict serialized account
         """
         profile = user.profile
+        has_accomplishments = settings.FEATURES.get('ENABLE_OPENBADGES') or False
```
Shouldn't `has_accomplishments` actually return whether or not the user has acquired badges?

Kelketek replied:
No, see other note. This field should probably be renamed instead.

Kelketek commented:
@talbs This PR has been updated with pagination controls. Please look over them and make comments as needed. I cloned the batman badge several times to fill up enough to have a few pages. Sorry for not being a bit more creative :)

@clrux I added a blank alt attribute to the badge images and the pagination changes have allowed for information about how many items are going to be listed.

I'm not able to split up the badge description information-- the text given is autogenerated as part of the legacy badge generation code. It might make sense to change how the generation of that text is done, but that's outside of the scope of this PR.

nasthagiri commented on openedx/core/djangoapps/user_api/accounts/views.py:
```diff
@@ -113,6 +114,13 @@ class AccountView(APIView):
 
+        **Params for GET**
+
+            * exclude: A field not to be returned in the results. This might
```
No longer need the `exclude` parameter.

nasthagiri commented on openedx/core/djangoapps/user_api/accounts/views.py:
```diff
@@ -98,6 +97,8 @@ class AccountView(APIView):
             * year_of_birth: The year the user was born, as an integer, or null.
             * account_privacy: The user's setting for sharing her personal
               profile. Possible values are "all_users" or "private".
+            * has_accomplishments: Signals whether badges are enabled on the
```
To support the UI requirements, `has_accomplishments` should indicate whether or not the user has actually earned any badges.

Kelketek replied:
This is causing problems. I shall change this field to read 'accomplishments_shared' so that it is more semantically correct. Setting this false for an empty accomplishment set, going way out of our way to load the accomplishments tab, inject an empty collection, and try hoping that no Backbone event attempts to reload the collection is hacky at best. This indicator should be for whether or not the accomplishment tab is loaded, not whether or not they have badges, for the sake of maintainable code.

nasthagiri replied:
Hmm.. The way I had interpreted the mobile UI requirements was that the "accomplishments" tab would appear only if badging was enabled, permissions allowed, *and* badges actually existed.  That is what we had written in our ticket: https://openedx.atlassian.net/browse/MA-1670.  However, this may have been a misinterpretation of the requirements.  And, if it over-complicates the code, it may not be worth it.  
FYI @explorerleslie.  Please note that we will show the accomplishments tab even if the user has no badges.  In the long run, this may be good, since we can entice the user to earn badges on an empty screen.

nasthagiri commented on openedx/core/lib/api/paginators.py:
```diff
@@ -15,6 +15,7 @@ class DefaultPagination(pagination.PageNumberPagination):
     by any subclass of Django Rest Framework's generic API views.
     """
     page_size_query_param = "page_size"
+    max_page_size = 100
```
Why did the default page_size for all paginators in the platform change?

Kelketek replied:
@nasthagiri The default page size did not change. I put a cap on the maximum size page a client can request. This is plugging a hole. We have 'page_size_query_param' set, but not 'max_page_size'. The REST framework documentation states that if you want to give a client the ability to specify their page size, you must specify a parameter name with 'page_size_query_param'. You then are to specify the max size that should be allowed. We never did that, so, in theory, one can pull arbitrarily large pages. The paginators we use handle ten items at a time, but I figured there might be occasions I missed where someone used more. 100 seemed like a reasonable cap.

nasthagiri replied:
Gotcha.  Ok.  I hope none of our clients are relying on a larger page size.

cahrens replied:
I don't think they are-- we generally use much smaller page sizes.

nasthagiri replied:
I meant external clients of our APIs.

nasthagiri commented on lms/static/js/student_profile/views/learner_profile_factory.js:
```diff
@@ -121,6 +125,15 @@
             ];
 
+            var badgeCollection = new PagingCollection();
+            badgeCollection.url = options.badges_api_url || "/dummy/url";
```
'Is `"/dummy/url"` intentional?

Kelketek replied:
@nasthagiri Yes. The collection won't initialize without a URL, but if we don't have one, we won't ever use it, because it means badges are disabled.

Kelketek replied:
Wait, perhaps not. Let me check something here...

nasthagiri commented on lms/templates/student_profile/badge_placeholder.underscore:
```diff
@@ -0,0 +1,10 @@
+    <p class="badge-description"><%- gettext('Start working toward your next learning goal.') %></p>
+    <a class="find-course" href="<%= find_courses_url %>"><span class="find-button-container"><%- gettext('Find a course') %></span></a>
+  </div>
+</div>
```
newline

nasthagiri commented on lms/static/sass/views/_learner-profile.scss:
```diff
@@ -307,8 +308,244 @@
+        padding-top: $baseline;
+    }
+
+    .pagination {
```
Is there no standard pagination framework used in the UI already?

Kelketek replied:
@nasthagiri A pagination framework exists, but it is not usable here. The SASS only exists in the CMS, and my attempts to move it to a common space proved fruitless. This is a modified copy of what is there.

cahrens replied:
I thought that there was some pagination code in common, though perhaps the sass is separate. @andy-armstrong, what do you think of this?

talbs replied:
I'd agree with some of the remarks above - copying this code from Studio is understandable, but I would not like to have multiple versions of this UI styling pop up in the LMS side of things. I'm more comfortable if:

1. You abstract and move this set of styling to be its own partial or within a more universal components area within the LMS sass space
2. Or, less ideally, make a note note that this styling needs to be abstracted and should not be copied to other Sass files alongside a clean up story/issue to abstract this out.

talbs replied:
Also, has this pagination pattern been reviewed from an a11y PoV by @cptvitamin? I don't believe all of Studio's UI patterns have in the past.

andy-armstrong replied:
We already had pagination in the LMS but it was scoped too tightly to teams. @frrrances made it more available here:

https://github.com/edx/edx-platform/pull/10997

nasthagiri commented on lms/static/js/student_profile/views/learner_profile_view.js:
```diff
@@ -26,6 +34,17 @@
                     ownProfile: this.options.ownProfile,
                     showFullProfile: this.showFullProfile()
                 }));
+                if (this.showFullProfile() && (this.options.accountSettingsModel.get('accomplishments_shared'))
```
Nit Question: is the call to `showFullProfile` redundant since the `accomplishments_shared` field would be set only if the profile is fully shared?

nasthagiri commented on openedx/core/djangoapps/user_api/urls.py:
```diff
@@ -3,7 +3,6 @@
 """
 
 from django.conf.urls import patterns, url
-
```
Nit: can revert change in this file.

nasthagiri commented:
Overall, this looks good to me.  Just waiting to see a green check from Jenkins and to check on code coverage.

Will you have a front-end developer look through it?  And we FYI-ed @cahrens so she's aware of the refactoring of the profile code and the duplication of the pagination Sass.

Kelketek commented:
@cahrens Moved the code back into the api, as you asked. @nasthagiri addressed your latest wave of notes.

cahrens commented on openedx/core/lib/api/permissions.py:
```diff
@@ -80,3 +86,64 @@ def has_object_permission(self, request, view, obj):
                 request.method in permissions.SAFE_METHODS)
+
+
+def get_profile_visibility(user_profile, user, configuration=None):
```
Can you explain more about what the problem is with putting these methods in the user_api api.py file? I really feel that they belong within the user_api package.

Kelketek replied:
@cahrens Let me attempt to do this one more time, and if it fails tell you why in more detail. Perhaps I did it wrong the first time.

cahrens replied:
The methods were originally in openedx/core/djangoapps/user_api/accounts/serializers.py. If they are exposed outside of user_api, they should be moved into openedx/core/djangoapps/user_api/accounts/api.py.

Kelketek replied:
@cahrens api imports from the other modules. Placing these functions in api would mean serializers has to import from a module that imports from it.

Kelketek replied:
@cahrens I am presently on hipchat if you need to discuss this further. I'd like to make sure I resolve your notes tonight if possible.

cahrens replied:
How about you leave the methods in user_api/accounts/serializers.py and put a wrapper for visible_fields in user_api/accounts/api.py (simply calls the implementation in serializers) for other djangoapps to use?

cahrens replied:
I'm heading to bed. If my suggestion doesn't work, we can pick up tomorrow.

Kelketek replied:
@cahrens I just finished attempting this, and it also failed, for similar reasons. When permissions attempts to load the function from user_api.accounts_api, it fails because the api module depends on a module which depends on the permissions module. I haven't figured out which one yet, and it would take more time than we have.

Kelketek replied:
@cahrens I found a way to do it. See the latest commit.

cahrens replied:
I think there is less urgency now (having checked with @smagoun). I'd like to have someone from TNL review this PR, and take time to figure out where these methods should go.

I created https://openedx.atlassian.net/browse/TNL-3921 for the TNL review.

Kelketek commented:
jenkins run js

cahrens commented:
Here is the TNL code review ticket-- https://openedx.atlassian.net/browse/TNL-3921. We hope to get to it on Thursday at the latest.

clrux commented:
@Kelketek Looking better. I'm still seeing some of the images without blank `alt` attributes. 
<img width="539" alt="screen shot 2015-12-15 at 10 56 41 am" src="https://cloud.githubusercontent.com/assets/2112024/11815651/a2848e14-a31a-11e5-9b47-2b02c6f3f475.png"> Specifically the "Robin Award".

As for the total count above the badges, what about including the keyword "accomplishments" or "badges" in case the student forgets that they're on the Accomplishments tab? So, "1 - 10 out of 33 total badges/accomplishments" or something like that.

Kelketek commented:
@clrux Could you check again on the alt images? If I check the source of the robin image in my browser, it shows the alt tab.

As for the message, we're using a pagination component which is shared across a few different views. I'd rather not change this since it's going to require updating all of those view and updating them to support internationalized plural/singulars of the things they're listing.

clrux commented:
@Kelketek I was looking at the sandbox. Was it updated, or is it stale? Looking at the code it looks like all of the `alt` attributes for the badge images have been nulled, which is good.

On @talbs' point, the pagination looks to be using pagination from Teams. I see "Teams Pagination" in the `aria-label`. Please update this to "Accomplishments pagination" or similar.

As for the pagination in general, it's functional but not ideal. This is probably outside the scope of this PR though, but the text input needs a label and that label needs to indicate the total number of pages. As it is now, one can enter a number (page to jump to) but have no idea how many pages there are total.

Kelketek commented:
@talbs I've moved the sass out in a place where it can be extended and used. @catong I've updated the text in the admin according to our meeting.

Could you both take a look?

Kelketek commented:
@clrux I've pushed an update which makes it say 'Accomplishments pagination' instead, and managed to update the input label to indicate out of how many pages the user is able to pick.

talbs commented:
FYI, @Kelketek - @frrrances just opened up this PR to abstract out pagination styling - https://github.com/edx/edx-platform/pull/10997. Its under review now, but my hopes is that you can wrap this into this UI work (pre or post-merge).

Kelketek commented:
jenkins run bokchoy

Kelketek commented:
@talbs Since this is merging to a feature branch, that PR shouldn't hold up this one. Chances are it will be merged before mine, so we can reconcile in the coming rebase.

talbs commented:
> @talbs Since this is merging to a feature branch, that PR shouldn't hold up this one. Chances are it will be merged before mine, so we can reconcile in the coming rebase.

@Kelketek, word. I just reviewed that other PR and things are looking good. I'm comfortable with the timing of the re-factor here coming before or after this merges to the feature branch.

clrux commented:
:+1:

Kelketek commented:
@nasthagiri Do I have your +1 on this? The T&L team still needs a look but I haven't seen any new comments from you.

nasthagiri commented:
@Kelketek yes.  :+1:

andy-armstrong commented on common/static/common/js/components/collections/paging_collection.js:
```diff
@@ -25,8 +25,31 @@
+
+                this.paginator_core = {
+                    type: 'GET',
+                        dataType: 'json',
```
Nit: the indentation is off here

andy-armstrong commented on common/static/common/js/components/collections/paging_collection.js:
```diff
@@ -25,8 +25,31 @@
+                        dataType: 'json',
+                        url: function () { return this.url; }
+                };
+                var self = this;
```
Nit: Put all variable definitions at the start of the function. This will be linted against shortly when we add JSCS to the platform.

http://jscs.info/rule/requireVarDeclFirst

andy-armstrong commented on common/static/common/js/components/collections/paging_collection.js:
```diff
@@ -25,8 +25,31 @@
+                };
             },
 
+            self: this,
```
Nit: what is this for? It is very confusing.

Kelketek replied:
@andy-armstrong `this` is context-dependent. If I want to reference the outer scope's `this` instead of the innter-scope's `this`, I have to give it a name. The name `self` is typically used in our backbone views when such a case arises.

Kelketek replied:
@andy-armstrong Oh, sorry, wrong instance of 'self'. That was a failed attempt to do what I was doing in initialization. I'll remove it.

andy-armstrong commented on common/static/common/js/components/views/list.js:
```diff
@@ -24,18 +24,24 @@
             },
 
+            renderCollection: function() {
+                // Render every item in the collection.
```
Nit: I'd make this a JSDoc comment for the function, rather than an inline comment.

andy-armstrong commented on lms/djangoapps/teams/static/teams/js/views/teams.js:
```diff
@@ -15,6 +15,8 @@
                 text: gettext('All teams')
             },
 
+            paginationLabel: gettext('Teams pagination'),
```
This was ``"Teams Pagination"`` before so I'd leave it as is unless this was a conscious decision to switch it.

andy-armstrong commented on lms/djangoapps/teams/views.py:
```diff
@@ -77,37 +77,12 @@ def team_post_save_callback(sender, instance, **kwargs):  # pylint: disable=unus
-
-
-class TopicsPagination(TeamAPIPagination):
+class TopicsPagination(DefaultPagination):
```
Sweet! Thanks for fixing this tech debt. FYI, here's the ticket where we've been tracking it:

https://openedx.atlassian.net/browse/TNL-3266

andy-armstrong commented:
I find the hover styling a little confusing for the tabs. It ends up underlining both tabs and making a single solid line.

![image](https://cloud.githubusercontent.com/assets/5985072/11847846/6655de56-a3ed-11e5-810a-4890b4e72c4e.png)

I tried the same thing on Teams and the styling is subtly different (although still not great). Is there any reason why we not using the same pattern?

![image](https://cloud.githubusercontent.com/assets/5985072/11847955/e2817012-a3ed-11e5-9ed6-f316c35ea715.png)

andy-armstrong commented:
I'd like to be able to send a link (or bookmark) the "Accomplishments" tab. Can it use the same browser history logic as the Teams tab?

andy-armstrong commented:
Is it intentional that the badge images shrink as I make the window narrower? I also find it odd that the badges can wrap in different ways depending upon the text length. For example, here some of the titles are wrapping beneath the image while some aren't.

![image](https://cloud.githubusercontent.com/assets/5985072/11848115/ccb1eb62-a3ee-11e5-9978-164f01690671.png)

andy-armstrong commented:
I'm a little uncomfortable with accomplishments being listed even for the user's limited profile. I thought that the intention was that the user opted in to sharing, but it seems that there is currently no way to choose to keep your badges private. Could we only show them for the full profile for now until we provide a more explicit option to the user.

cahrens commented:
I definitely agree with @andy-armstrong-- there has to be a way to hide badges for courses that you wish to be private. In the absence of fine-grain control on the badges, it makes sense to at least offer the "limited profile" heavy hammer approach (unless there is some way through a separate app to choose which badges are available to the API?).

andy-armstrong commented on lms/static/js/spec/student_profile/learner_profile_factory_spec.js:
```diff
@@ -62,6 +64,148 @@ define(['backbone', 'jquery', 'underscore', 'common/js/spec_helpers/ajax_helpers
+                var context = createProfilePage(true, {accomplishments_shared: true}),
+                    modeToggleView = context.learnerProfileView.modeToggleView;
+
+                requests.currentIndex = 1;
```
``currentIndex`` is intended to be an internal implementation detail, so you shouldn't be setting it. It also means that there is an unhandled request in the queue, which means that the code might not be behaving correctly. It would be preferable to handle the request in question.

I like to have the idiom of first verifying that the request was as expected, and then following up immediately with sending a mock response to that request. Something like this:

                AjaxHelpers.expectRequest(requests, 'GET', '/api/team/v0/topics/no_such_topic,course/1', null);
                AjaxHelpers.respondWithError(requests, 404);

https://github.com/edx/edx-platform/blob/master/lms/djangoapps/teams/static/teams/js/spec/views/teams_tab_spec.js#L86

This both consumes the request from the stack, and verifies that it was as expected.

andy-armstrong commented on lms/static/js/spec/student_profile/learner_profile_factory_spec.js:
```diff
@@ -62,6 +64,148 @@ define(['backbone', 'jquery', 'underscore', 'common/js/spec_helpers/ajax_helpers
+                AjaxHelpers.respondWithJson(requests, LearnerProfileHelpers.firstPageBadges);
+                LearnerProfileHelpers.expectPage(badgeListContainer, LearnerProfileHelpers.firstPageBadges);
+            });
+
```
Add a test that verifies what happens when the badge API returns an error.

andy-armstrong commented on lms/static/js/student_profile/views/badge_list_container.js:
```diff
@@ -0,0 +1,23 @@
+
+            return BadgeListContainer;
+        });
+}).call(this, define || RequireJS.define);
```
Nit: missing newline

andy-armstrong commented on lms/static/js/student_profile/views/badge_list_container.js:
```diff
@@ -0,0 +1,23 @@
+                listViewClass: BadgeListView,
+                viewTemplate: BadgeListTemplate,
+                isZeroIndexed: true,
+                paginationLabel: gettext("Accomplishments pagination")
```
Nit: this should use title case, i.e. "Accomplishments Pagination"

andy-armstrong commented on lms/static/sass/views/_learner-profile.scss:
```diff
@@ -307,8 +308,123 @@
         }
     }
-}
+}
```
Nit: missing newline

andy-armstrong commented on lms/templates/student_profile/badge.underscore:
```diff
@@ -0,0 +1,15 @@
+  <div class="badge-date-stamp">
+    <%-
+      interpolate(
+        gettext('Earned %(created)s.'),
```
This will be hard to translate, so add a translation comment describing where it is used.

andy-armstrong commented:
@Kelketek This is an excellent feature, and I love how it is all built using reusable components. I have a few nits and some suggestions for the unit tests, but otherwise this looks great. Let me know when it is ready for my second review.

catong commented on lms/djangoapps/badges/migrations/0003_schema__add_event_configuration.py:
```diff
@@ -20,9 +20,9 @@ class Migration(migrations.Migration):
-                ('courses_completed', models.TextField(default=b'', help_text="On each line, put the number of completed courses to award a badge for, a comma, and the slug of a badge class you have created with the issuing component 'edx__course'. For example: 3,course-v1:edx/Demo/DemoX", blank=True)),
-                ('courses_enrolled', models.TextField(default=b'', help_text="On each line, put the number of enrolled courses to award a badge for, a comma, and the slug of a badge class you have created with the issuing component 'edx__course'. For example: 3,course-v1:edx/Demo/DemoX", blank=True)),
-                ('course_groups', models.TextField(default=b'', help_text="On each line, put the slug of a badge class you have created with the issuing component 'edx__course' to award, a comma, and a comma separated list of course keys that the user will need to complete to get this badge. For example: slug_for_compsci_courses_group_badge,course-v1:CompSci+Course+First,course-v1:CompsSci+Course ...
+                ('courses_completed', models.TextField(default=b'', help_text="On each line, put the number of completed courses to award a badge for, a comma, and the slug of a badge class you have created with the issuing component 'edx__course'. For example: 3,hat_trick", blank=True)),
```
@Kelketek I recommend using something that global audiences will understand in the example: "hat_trick" -> "completed_3_courses".

catong replied:
@Kelketek Can you please confirm whether multiple lines are allowed? It sounds that way from the help text.

catong commented on lms/djangoapps/badges/migrations/0003_schema__add_event_configuration.py:
```diff
@@ -20,9 +20,9 @@ class Migration(migrations.Migration):
-                ('courses_enrolled', models.TextField(default=b'', help_text="On each line, put the number of enrolled courses to award a badge for, a comma, and the slug of a badge class you have created with the issuing component 'edx__course'. For example: 3,course-v1:edx/Demo/DemoX", blank=True)),
-                ('course_groups', models.TextField(default=b'', help_text="On each line, put the slug of a badge class you have created with the issuing component 'edx__course' to award, a comma, and a comma separated list of course keys that the user will need to complete to get this badge. For example: slug_for_compsci_courses_group_badge,course-v1:CompSci+Course+First,course-v1:CompsSci+Course ...
+                ('courses_completed', models.TextField(default=b'', help_text="On each line, put the number of completed courses to award a badge for, a comma, and the slug of a badge class you have created with the issuing component 'edx__course'. For example: 3,hat_trick", blank=True)),
+                ('courses_enrolled', models.TextField(default=b'', help_text="On each line, put the number of enrolled courses to award a badge for, a comma, and the slug of a badge class you have created with the issuing component 'edx__course'. For example: 3,hat_trick", blank=True)),
```
@Kelketek I recommend using something that global audiences will understand in the example: "hat_trick" -> "enrolled_3_courses".

catong commented on lms/djangoapps/badges/migrations/0003_schema__add_event_configuration.py:
```diff
@@ -20,9 +20,9 @@ class Migration(migrations.Migration):
-                ('course_groups', models.TextField(default=b'', help_text="On each line, put the slug of a badge class you have created with the issuing component 'edx__course' to award, a comma, and a comma separated list of course keys that the user will need to complete to get this badge. For example: slug_for_compsci_courses_group_badge,course-v1:CompSci+Course+First,course-v1:CompsSci+Course ...
+                ('courses_completed', models.TextField(default=b'', help_text="On each line, put the number of completed courses to award a badge for, a comma, and the slug of a badge class you have created with the issuing component 'edx__course'. For example: 3,hat_trick", blank=True)),
+                ('courses_enrolled', models.TextField(default=b'', help_text="On each line, put the number of enrolled courses to award a badge for, a comma, and the slug of a badge class you have created with the issuing component 'edx__course'. For example: 3,hat_trick", blank=True)),
+                ('course_groups', models.TextField(default=b'', help_text="Each line is a comma-separated list. The first item in each line is the slug of a badge class to award, with an issuing component of 'edx__course'. The remaining items in each line are the course keys the user will need to complete to get the badge. For example: slug_for_compsci_courses_group_badge,course-v1:CompSci+Course ...
```
"course keys the user will need to complete" -> "course keys for the courses that learners need to complete"

"to get the badge" -> "to be awarded the badge".

catong replied:
@Kelketek Can you pls confirm the format that course keys need to be in? Is there any validation or check?

Kelketek replied:
@catong It uses CourseKey.from_string to resolve what they input into a course string. So any course identifier which can be resolved that way should be fine, and it will give an error if they didn't put in a valid one.

catong commented on lms/djangoapps/badges/models.py:
```diff
@@ -216,7 +217,7 @@ class CourseEventBadgesConfiguration(ConfigurationModel):
             u"On each line, put the number of completed courses to award a badge for, a comma, and the slug of a "
             u"badge class you have created with the issuing component 'edx__course'. "
-            u"For example: 3,course-v1:edx/Demo/DemoX"
+            u"For example: 3,hat_trick"
```
@Kelketek Please use something that global audiences will understand in the example: "hat_trick" -> "completed_3_courses".

catong commented on lms/djangoapps/badges/models.py:
```diff
@@ -224,7 +225,7 @@ class CourseEventBadgesConfiguration(ConfigurationModel):
             u"On each line, put the number of enrolled courses to award a badge for, a comma, and the slug of a "
             u"badge class you have created with the issuing component 'edx__course'. "
-            u"For example: 3,course-v1:edx/Demo/DemoX"
+            u"For example: 3,hat_trick"
```
@Kelketek Please use something that global audiences will understand in the example: "hat_trick" -> "enrolled_3_courses".

cahrens commented on openedx/core/lib/api/permissions.py:
```diff
@@ -80,3 +82,31 @@ def has_object_permission(self, request, view, obj):
                 request.method in permissions.SAFE_METHODS)
+
+
+def is_field_shared_factory(field_name):
```
@andy-armstrong and I talked this over-- we feel that the best thing is to move this permissions class into the user_api package itself, since it is specific to user_api (whereas this core api package is meant to hold more generic classes). You can create a permissions.py file within user_api/accounts to hold it. Hopefully that will break the circular dependency issue.

Kelketek commented:
@andy-armstrong I'm working on your notes. I wasn't aware of the Teams project and the components it added until after most of the reviews were completed-- when I was asked to paginate the badging results, when I ran across it and had to patch up the pagination component to make it more generally usable.

The tabbed view component I've just now discovered as of a few moments ago. Looks like my ModeToggle view is redundant. It's going to take time to refactor everything to use the existing view (which includes history and, thus, the ability to link directly to badges). I'll be pushing fixes for the other issues first, then I suppose I'll get to redoing a bunch of stuff.

talbs commented:
@Kelketek, https://github.com/edx/edx-platform/pull/10997 has been merged. Can you refactor your styling/markup to leverage this new app-wide pagination element? Let me know the timing on that as well as when I can help confirm that you've removed any of the Sass you've had to write/port yourself.

Thanks much!

andy-armstrong commented:
@Kelketek One thing I forgot to mention yesterday: should the badges be ordered reverse chronologically? It seems odd to show my oldest badges first, especially if I've been using edX for a number of years. I think it would make sense to show the most recent ones first.

Kelketek commented:
I've had to rebase this to get @talbs 's request of incorperating the recently merged pagination SASS chances done. 

@catong You will find the updates you requested in https://github.com/edx/edx-platform/commit/bfc1956e08908d3bc779cf54940de0b5beb32234

120f1cb Is the start of the most recent wave of changes. It took all day to refactor the views to use the Team's tabbed view (I rather wish I had known about that beforehand instead of just before we were done). I will not be able to make the badges reflow quite as nicely as we'd like-- we don't have a proper grid system in the LMS that can be used consistently and we do not have time to implement one now-- in fact we really don't have time to do much else with this PR.

@andy-armstrong Please give this another look.

Kelketek commented:
jenkins run lettuce

Kelketek commented:
In the interest of expediency, the sandbox has been updated with the share modal for https://github.com/edx/edx-platform/pull/11028 . It includes all of the current work here, though.

Kelketek commented:
jenkins run bok_choy

Kelketek commented:
jenkins run bokchoy

andy-armstrong commented:
@Kelketek Thanks so much for the extensive reworking, and sorry you weren't aware of some of these features until late in the day. :+1: from me once the hard-coded "edX" string has been removed.

Kelketek commented:
jenkins run js

talbs commented:
@Kelketek, you still have my :+1: for this. Thanks for the extra syncing up with our recent changes.

edx-webhook commented:
Thanks for the pull request, @Kelketek! I've created [OSPR-1567](https://openedx.atlassian.net/browse/OSPR-1567) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.
